Update wasm threads proposal tests and fix cmpxchg_u (edited)

eugeneia opened the pull request:
#### b2062fb65324b2f5b903e9762e061533b48b26f8
<pre>
Update wasm threads proposal tests and fix cmpxchg_u (edited)
<a href="https://bugs.webkit.org/show_bug.cgi?id=263292">https://bugs.webkit.org/show_bug.cgi?id=263292</a>
<a href="https://rdar.apple.com/117102231">rdar://117102231</a>

Reviewed by Justin Michaud and Yusuke Suzuki.

In <a href="https://github.com/WebAssembly/threads/issues/195">https://github.com/WebAssembly/threads/issues/195</a>, it was documented
that JSC implements cmpxchg_u without wrapping the expected value.

Previously, the spec tests and reference interpreter followed this interpretation,
but the spec text, SpiderMonkey and v8 did the wrapping.

The spec tests were updated, and this new behaviour is simpler anyway,
so let&apos;s just do it.

* JSTests/wasm/spec-tests/atomic.wast.js: Added.
* JSTests/wasm/spec-tests/binary-leb128.wast.js:
* JSTests/wasm/spec-tests/binary.wast.js:
* JSTests/wasm/spec-tests/br_table.wast.js:
* JSTests/wasm/spec-tests/elem.wast.js:
* JSTests/wasm/spec-tests/float_literals.wast.js:
* JSTests/wasm/threads-spec-tests/atomic-signed.wast.js:
* Source/JavaScriptCore/llint/WebAssembly32_64.asm:
* Source/JavaScriptCore/llint/WebAssembly64.asm:
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::emitAtomicCompareExchange):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::emitAtomicCompareExchange):

Canonical link: <a href="https://commits.webkit.org/270828@main">https://commits.webkit.org/270828@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4000a748f000e3b81e562b277604e6bf10a7c1f3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26363 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4976 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27625 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28471 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24133 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6775 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2405 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24121 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26623 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3829 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22670 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3392 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3418 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23642 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29061 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/22988 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24043 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24041 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29704 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25579 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3446 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1667 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27597 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4869 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33024 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6368 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3924 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7138 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3770 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (approved):
r=me

Constellation reviewed (approved):
r=me
Can you also update InPlaceInterpreter.asm?

eugeneia committed:
Update wasm threads proposal tests and fix cmpxchg_u (edited)
https://bugs.webkit.org/show_bug.cgi?id=263292
rdar://117102231

Reviewed by Justin Michaud and Yusuke Suzuki.

In https://github.com/WebAssembly/threads/issues/195, it was documented
that JSC implements cmpxchg_u without wrapping the expected value.

Previously, the spec tests and reference interpreter followed this interpretation,
but the spec text, SpiderMonkey and v8 did the wrapping.

The spec tests were updated, and this new behaviour is simpler anyway,
so let's just do it.

* JSTests/wasm/spec-tests/atomic.wast.js: Added.
* JSTests/wasm/spec-tests/binary-leb128.wast.js:
* JSTests/wasm/spec-tests/binary.wast.js:
* JSTests/wasm/spec-tests/br_table.wast.js:
* JSTests/wasm/spec-tests/elem.wast.js:
* JSTests/wasm/spec-tests/float_literals.wast.js:
* JSTests/wasm/threads-spec-tests/atomic-signed.wast.js:
* Source/JavaScriptCore/llint/WebAssembly32_64.asm:
* Source/JavaScriptCore/llint/WebAssembly64.asm:
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::emitAtomicCompareExchange):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::emitAtomicCompareExchange):

Canonical link: https://commits.webkit.org/270828@main
